byteps/server/server.cc

```diff
@@ -23,48 +24,51 @@ using namespace ps;
 
 // engine related
 std::vector<PriorityQueue*> engine_queues_;
-std::vector<std::thread *> engine_threads_;
+std::vector<std::thread*> engine_threads_;
 
 BytePSArray* GetStore(uint64_t key) {
   std::lock_guard<std::mutex> lock(store_mu_);
   return &store_[key];
 }
 
-void SendPushResponse(uint64_t key, const ps::KVMeta& req, ps::KVServer<char>* server){
+void SendPushResponse(uint64_t key, const ps::KVMeta& req,
+                      ps::KVServer<char>* server) {
   auto iterator = push_response_map_.find(key);
-  if (iterator == push_response_map_.end()) { // new key
+  if (iterator == push_response_map_.end()) {  // new key
     ps::KVPairs<char> response;
-    push_response_map_[key] = response; // add to the map
+    push_response_map_[key] = response;  // add to the map
     server->Response(req, response);
-  } else { // not new key, then reuse the memory address to avoid ibv_reg_mr on RDMA data path
-    ps::KVPairs<char> *response = &iterator->second;
+  } else {  // not new key, then reuse the memory address to avoid ibv_reg_mr on
+            // RDMA data path
+    ps::KVPairs<char>* response = &iterator->second;
     server->Response(req, *response);
   }
 }
 
-void SendPullResponse(const DataHandleType type,
-                      const uint64_t key,
-                      const ps::KVMeta& req_meta,
-                      ps::KVServer<char>* server) {
+void SendPullResponse(const DataHandleType type, const uint64_t key,
+                      const ps::KVMeta& req_meta, ps::KVServer<char>* server) {
   std::lock_guard<std::mutex> lock(pullresp_mu_);
-  auto stored = GetStore(key);
-  CHECK(stored->tensor) << "init " << key << " first";
-  auto len = stored->len;
+  auto& updates = update_buf_[key];
+  CHECK(updates.merged.tensor) << "init " << key << " first";
+  char* data = updates.merged.tensor;
+  auto len = updates.merged.len;
 
   // send pull response
   auto iterator = pull_response_map_.find(key);
-  if (iterator == pull_response_map_.end()) { // new key
+  if (iterator == pull_response_map_.end()) {  // new key
     ps::KVPairs<char> response;
     response.keys = {EncodeKey(key)};
     response.lens = {len};
-    response.vals = ps::SArray<char>(stored->tensor, len, false); // zero copy
-    pull_response_map_[key] = response; // add to the map
+    response.vals = ps::SArray<char>(data, len, false);  // zero copy
+    pull_response_map_[key] = response;                  // add to the map
```

Comment:
This is hard for me to understand. Why send the update_buf_ rather than stored_ here?

Reply:
Some tensors are compressed but others not. Compressors never change  input tensors and always return a new pointer pointing to an internal buffer where the compressed tensors are located. But for those uncompressed tensors, their data are stored in `stored_`. For convenience, `update_buf_` serves as a unified tensor waiting to send.